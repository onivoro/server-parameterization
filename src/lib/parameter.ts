export abstract class Parameter<T> {
    key() {
        return this.transform(this.id());
    }

    abstract value(): T;
    protected abstract id(): string;
    protected abstract transform(id: string): string;
}