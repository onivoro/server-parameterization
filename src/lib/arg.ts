import * as kebabcase from 'lodash.kebabcase';
import { Parameter } from './parameter';
import * as minimist from 'minimist';

export abstract class Arg<T> extends Parameter<T> {
    protected transform(id: string): string {
        return kebabcase(id).toLowerCase();
    }

    value() {
        return minimist(process.argv.slice(2), { default: {} })[this.key()];
    }
}
