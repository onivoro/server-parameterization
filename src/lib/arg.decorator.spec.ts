import "reflect-metadata";
import { Parameter } from './parameter.class';
import { Arg } from './arg.decorator';

let minimistMock = jest.fn().mockReturnValue({
    'some-string': 'asdf',
    'some-number': '337',
    'some-bool': 'false',
    'other-bool': 'true',
    'some-date': '2020-12-20T11:31:41.378Z',
});

@Arg(String, minimistMock) export class SomeString extends Parameter<string> { }
@Arg(Number, minimistMock) export class SomeNumber extends Parameter<number> { }
@Arg(Boolean, minimistMock) export class SomeBool extends Parameter<boolean> { }
@Arg(Boolean, minimistMock) export class OtherBool extends Parameter<boolean> { }
@Arg(Date, minimistMock) export class SomeDate extends Parameter<Date> { }

describe('argDecorator', () => {

    it.each([
        new SomeString(),
        new SomeNumber(),
        new SomeBool(),
        new OtherBool(),
        new SomeDate()
    ])('reads the conventionally-implied process argument and stores both raw and typed representations thereof', (argVar) => {
        expect(argVar).toMatchSnapshot();
    });
});
