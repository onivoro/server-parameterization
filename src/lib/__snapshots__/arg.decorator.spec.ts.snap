// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`argDecorator reads the conventionally-implied process argument and stores both raw and typed representations thereof 1`] = `
SomeString {
  "key": "some-string",
  "source": "process.argv",
  "typedValue": "asdf",
  "value": "asdf",
}
`;

exports[`argDecorator reads the conventionally-implied process argument and stores both raw and typed representations thereof 2`] = `
SomeNumber {
  "key": "some-number",
  "source": "process.argv",
  "typedValue": 337,
  "value": "337",
}
`;

exports[`argDecorator reads the conventionally-implied process argument and stores both raw and typed representations thereof 3`] = `
SomeBool {
  "key": "some-bool",
  "source": "process.argv",
  "typedValue": false,
  "value": "false",
}
`;

exports[`argDecorator reads the conventionally-implied process argument and stores both raw and typed representations thereof 4`] = `
OtherBool {
  "key": "other-bool",
  "source": "process.argv",
  "typedValue": true,
  "value": "true",
}
`;

exports[`argDecorator reads the conventionally-implied process argument and stores both raw and typed representations thereof 5`] = `
SomeDate {
  "key": "some-date",
  "source": "process.argv",
  "typedValue": 2020-12-20T11:31:41.378Z,
  "value": "2020-12-20T11:31:41.378Z",
}
`;
